### DISCUSSION TITLE
plotting time-series-based posterior samples

### DISCUSSION THREAD

**Discussion by skaur42** [arviz-devs/arviz#2075] [Q&A] [answered]
Hi there,

I’m interested in plotting time-series-based posterior samples (plots of distributions) and am trying to figure out if it is appropriate for inclusion in ArViz and, if it is, how? [This](https://github.com/AmpersandTV/pymc3-hmm/blob/main/examples/poisson_time_varying_example.ipynb) is an example of the plotting feature in pymc3-hmm (vertically aligned histograms at each time). I want to visualize each distribution at each point in time, like one would if they plotted a standard time-series with only a single point at each time, and we want it to work for a large series (e.g. 10-100k time points and large sample sizes).

**Comment by OriolAbril** [comment, marked as answer]
If I understand correctly, this is already possible (but tricky and needs way more documentation!). ArviZ has some "full blown figures" but it also builds on top of xarray which already supports some plotting capabilities. Moreover, we recently released [xarray-einstats](https://einstats.python.arviz.org/en/latest/) which allows to easily generate histograms over a subset of the data dimensions.

I took a similar model (I think) I had at hand, the one from [this blog post I wrote recently](https://oriolabrilpla.cat/python/arviz/stan/xarray/xarray-einstats/2022/04/25/einstats-hmm-cmdstanpy.html). The model and posterior sampling use cmdstanpy, but after that we get an InferenceData, so it doesn't matter anymore which was the PPL used, once you get an InferenceData you should be able to use all the same.

I updated the last plot of the "Simple HMM Example" to the following code:

```
from xarray_einstats.numba import histogram

hist = histogram(idata.posterior_predictive["y"], dims=["chain", "draw"], bins=50)

_, ax = plt.subplots()
ax.plot(hmm_data["y"], "k-", zorder=3, label="Observed", alpha=.5)
ax.set_title("Observed vs Predicted Output")
ax.set_ylabel("Observation Value")
ax.set_xlabel("Time")

hist.plot.imshow(y="left_edges", x="time", ax=ax, cmap="Blues", alpha=0.7)
```

which generates what I believe is the plot you want:

![image](https://user-images.githubusercontent.com/23738400/180667014-817b7bc5-10dc-4aa5-a73b-b824fe1af878.png)

Useful references:
* https://einstats.python.arviz.org/en/latest/api/generated/xarray_einstats.numba.histogram.html (haven't used it yet, but it might also be worth looking into https://xhistogram.readthedocs.io/en/latest/index.html if you don't want the numba dependency or simply to compare)
* https://docs.xarray.dev/en/stable/generated/xarray.plot.imshow.html
